### ISSUE TITLE
Selected text not pasted into DeepL window/tab

### ISSUE THREAD

**Issue by DenB10** [itagagaki/DeepL-Selected-Text#6]
Hi itagagaki,
I think your add-on needs a fix: The text selected for translation appears to be transferred to DeepL but is not displayed there, nor is the translation.
Best regards,
DenB

**Comment by itagagaki** [comment]
Thank you for your report.
We apologize for the inconvenience, but could you please try the following

1. Go to "Settings" -> "Privacy and Security"
2. Click "Show Cookies..." under "Web Content"
3. Type "deepl" in the "Search" field
4. Remove all cookies match

If it does not improve once, try restarting Thunderbird and removing the cache several times.

**Comment by DenB10** [comment]
I'm the one who's sorry for the inconvenience, because today it's working again as expected.
It was probably a problem at DeepL???
Thank you very much for the responsiveness.
DenB

**Comment by itagagaki** [comment]
Glad it has been resolved.
Similar cases have been reported by others recently and I have experienced it as well. The real cause is still unknown.

**Comment by MarkyMarkDE** [comment]
did not solve my problem.

now i have much more error logs in console

15:15:35.656 target is null 2 ExtensionParent.jsm:98
15:15:35.657 target is null ExtensionParent.jsm:98
15:15:35.657 target is null 2 ExtensionParent.jsm:98
15:15:35.657 target is null ExtensionParent.jsm:98
15:15:35.658 sendRemoveListener on closed conduit languagetool-mailextension@languagetool.org.666 2 ConduitsChild.jsm:112
15:15:36.934 TypeError: r is null translator_late.min.$4e2dab.js:44:688844
15:15:43.709 sendRemoveListener on closed conduit languagetool-mailextension@languagetool.org.2886218022913 2 ConduitsChild.jsm:112
15:19:51.955 target is null 2 ExtensionParent.jsm:98
15:19:51.956 target is null ExtensionParent.jsm:98
15:19:51.956 target is null 2 ExtensionParent.jsm:98
15:19:51.956 target is null ExtensionParent.jsm:98
15:19:51.957 sendRemoveListener on closed conduit languagetool-mailextension@languagetool.org.687 2 ConduitsChild.jsm:112
15:19:53.173 TypeError: r is null translator_late.min.$4e2dab.js:44:688844
15:20:01.477 sendRemoveListener on closed conduit languagetool-mailextension@languagetool.org.3023656976385 2 ConduitsChild.jsm:112

**Comment by itagagaki** [comment]
I tried it on Ubuntu and reproduced the problem.
From the following error message output to the console, it seems that this is a problem with deepl.com.
Unfortunately, there is nothing we can do about it here.
We will have to wait for them to discover  and fix the problem themselves.
```
JavaScript error: https://static.deepl.com/js/translator_late.min.$4e2dab.js, line 44: TypeError: r is null
```

**Comment by MarkyMarkDE** [comment]
> I tried it on Ubuntu and reproduced the problem. From the following error message output to the console, it seems that this is a problem with deepl.com. Unfortunately, there is nothing we can do about it here. We will have to wait for them to discover and fix the problem themselves.
> 
> ```
> JavaScript error: https://static.deepl.com/js/translator_late.min.$4e2dab.js, line 44: TypeError: r is null
> ```

and why the firefox version of deepl work?

**Comment by itagagaki** [comment]
As a result of the various conditions, the server at DeepL may or may not pass thru this line 44, and if it does, whether or not r is null at that time. Whether the problem occurs or not depends on this.
Alternatively, since DeepL has a load balancing mechanism to cope with a large numbers of accesses, it is possible that this $4e2dab.js program itself may be slightly different depending on the server to which it is assigned.
The conditions could be user agent, cookie, IP, country, time, and various other coincidences.

**Comment by MarkyMarkDE** [comment]
do i read this well, that the firefox addon "to deepl" did not use the deepl API and instead they send the data to deepl.com/translator#null?
https://github.com/xpmn/firefox-to-deepl/blob/master/src/background.js

**Comment by itagagaki** [comment]
The following information is output to the Thunderbird error console. This cookie issue may result in the 'r is null' error.

```
Some cookies are misusing the recommended “SameSite“ attribute
Cookie “rgReload” has been rejected because it is already expired.
Cookie “userCountry” has been rejected because it is already expired.
Cookie “dapUid” has been rejected because it is already expired.
```

It is possible that this cookie problem is only occurs in Thunderbird and not in other browsers such as Firefox.
At some point, Thunderbird may have undergone a change in cookie handling.

**Comment by itagagaki** [comment]
However, the following report suggests that this is still a site issue.
https://bugzilla.mozilla.org/show_bug.cgi?id=1691960

**Comment by MarkyMarkDE** [comment]
> However, the following report suggests that this is still a site issue. https://bugzilla.mozilla.org/show_bug.cgi?id=1691960

do not think that this problem is caused by cookies.
Please read this: https://addons.mozilla.org/de/firefox/addon/to-deepl/reviews/
Seems that the API of deepl is since longer time broken, "to depl" Addon for firefox was also affect and solved this problem.

**Comment by itagagaki** [comment]
Maybe the reason firefox-to-deepl didn't work at one point is not the same as the reason DeepL Selected Text sometimes doesn't work now. From what I've seen in the code, the mechanics of both are the same. The difference is that they use null as the source language. I tried to mimic this, but it didn't solve the problem. DeepL Selected Text is not working because of a glitch in the server side (DeepL side) script. And I suspect that the glitch is caused by its misuse of cookies. I don't know why the problem only occurs in Thunderbird, but either other browsers allow the incorrect use of cookies, or the script is running under different conditions on the server due to differences in user agents containing browser information as mentioned above, or the scripts running at all may be different.

**Comment by itagagaki** [comment]
This add-on works in Thunderbird on Windows. It certainly did not work recently. However, I am not sure if the results of deleting cookies, etc. were successful or not, but it works anyway.
There are warnings about misuse of cookies on Windows as well as Linux. The fact that it still works means that their misuse of cookies may not actually be the root cause.
In any case, if there is a problem with the add-on code, it does not explain why it works with Thunderbird on Windows.

**Comment by FranzBias** [comment]
Hello itagagaki,
I have the same problem as @DenB10 (at the top).
I have uninstalled the plug-in and then reinstalled it, and also tried clearing the cache, but the problem persists.
I read that the error seems to be in the cookies of DeepL and not in your plug-in. 

My question is: has anyone reported this to DeepL yet? If not, I'll do it (I have a Pro subscription to DeepL, maybe they'll listen to me :X). In case I would need more information about the error and which cookie produces it (I'm not very familiar with these things, error codes, consoles, etc, and it would be helpful to explain the problem.

Greetings and thanks for this extension.

**Comment by itagagaki** [comment]
@FranzBias please do that.
Below is the Thunderbird error log.

```
Cookie "rgReload" does not have a proper "SameSite" attribute value. Soon, cookies without the "SameSite" attribute or with an invalid value will be treated as "Lax". This means that the cookie will no longer be sent in third-party contexts. If your application depends on this cookie being available in such contexts, please add the "SameSite=None" attribute to it. To know more about the "SameSite" attribute, read https://developer.mozilla.org/docs/Web/HTTP/Headers/Set-Cookie/SameSite
Cookie "userCountry" does not have a proper "SameSite" attribute value. Soon, cookies without the "SameSite" attribute or with an invalid value will be treated as "Lax". This means that the cookie will no longer be sent in third-party contexts. If your application depends on this cookie being available in such contexts, please add the "SameSite=None" attribute to it. To know more about the "SameSite" attribute, read https://developer.mozilla.org/docs/Web/HTTP/Headers/Set-Cookie/SameSite
Cookie "releaseGroups" does not have a proper "SameSite" attribute value. Soon, cookies without the "SameSite" attribute or with an invalid value will be treated as "Lax". This means that the cookie will no longer be sent in third-party contexts. If your application depends on this cookie being available in such contexts, please add the "SameSite=None" attribute to it. To know more about the "SameSite" attribute, read https://developer.mozilla.org/docs/Web/HTTP/Headers/Set-Cookie/SameSite
Cookie "rgReload" has been rejected because it is already expired. translator
Cookie "userCountry" has been rejected because it is already expired. translator
Cookie "releaseGroups" has been rejected because it is already expired. translator
TypeError: r is null translator_late.min.$2781ab.js:44:703206
```

Thank you for your cooperation.

**Comment by FranzBias** [comment]
@itagagaki 
Done! 
Hopefully they will take my request into consideration ;).
Greetings,
Franz

**Comment by MarkyMarkDE** [comment]
no further headaches needed, use this Addon, here is deepl in Thunderbird with a workarround fixed!
https://addons.thunderbird.net/en-US/thunderbird/addon/web_translate/
- also for firefox available -
https://addons.mozilla.org/en-US/firefox/addon/web_translate/

**Comment by itagagaki** [comment]
Thank you @FranzBias
We often find it difficult to have a conversation with a company's support department. They often prioritize getting complaints out of the way quickly over trying to understand the nature of the problem.

But I appreciate your efforts.

**Comment by itagagaki** [comment]
@MarkyMarkDE 
I believe that this is a bug in DeepL translator. It must be their degradation because I'm sure they didn't have this bug when Thunderbird and my add-on worked as expected.

It would be interesting to know how the add-on you referred me to worked around that bug in DeepL translator. But it seems that it is not open source, so I could not learn from the code history. I might be able to do reverse engineer the current code of the provided add-on and analyze it, but I don't want to do that because it's not something I'm very impressed with.

**Comment by DenB10** [comment]
Hello everyone,
The problem came back for me but I noticed that by deleting only the cookie named "releaseGroups", it works correctly again.
Unfortunately, you have to delete this cookie every time you run Thunderbird.
I just hope my remark can help find a workaround!

**Comment by itagagaki** [comment]
@MarkyMarkDE 
So he uses Vue.js. But that should have nothing to do with whether it is open source or not.. Anyway, he got around the bug by injecting the text into the text box on the DeepL page instead of passing it to the server with the GET method? I see.

I wouldn't want to do it that way. Because it depends on the design of the web form that DeepL generates. They can change it at their whim. I don't want to deal with their changes every time.

@FranzBias 
Sometimes it works, sometimes it doesn't. I assume that this is a result of the load balancing system of DeepL as I mentioned before, where requests are distributed to different servers and different scripts are run.
In other words, it is a matter of luck.

@DenB10 
That is very important information!
The measure to delete that cookie before making a request to DeepL seems to work (though I'm not sure if it's a perfect solution). I'll give it a try.

**Comment by DenB10** [comment]
To complete my comment:

To avoid having to delete the cookie each time Thunderbird is restarted, I have selected the following for cookies:
"Keep until: ask me every time"

**Comment by MarkyMarkDE** [comment]
> @MarkyMarkDE So he uses Vue.js. But that should have nothing to do with whether it is open source or not.. Anyway, he got around the bug by injecting the text into the text box on the DeepL page instead of passing it to the server with the GET method? I see.
> 
> I wouldn't want to do it that way. Because it depends on the design of the web form that DeepL generates. They can change it at their whim. I don't want to deal with their changes every time.
> 
> @FranzBias Sometimes it works, sometimes it doesn't. I assume that this is a result of the load balancing system of DeepL as I mentioned before, where requests are distributed to different servers and different scripts are run. In other words, it is a matter of luck.
> 
> @DenB10 That is very important information! The measure to delete that cookie before making a request to DeepL seems to work (though I'm not sure if it's a perfect solution). I'll give it a try.

in php i do pass it like that, maybe this can be a similar way for you?
german deepl UI: ``https://www.deepl.com/de/translator#en/de/'.htmlentities(substr(TEXT, 0, 1500), ENT_QUOTES, 'ISO-8859-15')``
english deepl UI: ``https://www.deepl.com/en/translator#de/en/'.htmlentities(substr(TEXT, 0, 1500), ENT_QUOTES, 'ISO-8859-15')``

and the URI is open in a new window with window.open method

**Comment by itagagaki** [comment]
Thank you all for your cooperation.
Release version 1.6.0 has been released in this repository.
I have also submitted it to ATN and should be able to update once it is approved.
If it still doesn't work after the update, try deleting the cookie manually and restarting once. After that I think it's fine.

**Comment by DenB10** [comment]
@itagagaki 
It still doesn't work quite right:
After restarting Thunderbird, the first time you request a translation, the window remains empty as before, and the offending releaseGroups cookie is not deleted. You have to request the translation a second time for it to work (the cookie is then correctly deleted).

**Comment by DenB10** [comment]
@itagagaki 
I think the line `messenger.cookies.remove({name: "releaseGroups", url: "https://www.deepl.com"});` in the file background.js doesn't work the first time because the releaseGroups cookie hasn't yet been created.
I think it might be necessary to delete the cookie after injecting the text and then refresh the display?

**Comment by itagagaki** [comment]
@DenB10 Ah, you are right. I'll try to improve it.

**Comment by DenB10** [comment]
> Unfortunately, to work around a DeepL bug, the Open in New Window option had to change from the previous popup window style to a normal window with UI.)

Anyway, it's working fine now, thanks.
Just one question: why is the change from 'popup' to 'normal' necessary for the window?

**Comment by itagagaki** [comment]
> Just one question: why is the change from 'popup' to 'normal' necessary for the window?

Because couldn't resolve an error "tabs.reload is not applicable to this tab."

**Comment by itagagaki** [comment]
> Because couldn't resolve an error "tabs.reload is not applicable to this tab."

This is reportedly fixed in Thunderbird 115. And in earlier versions it was solved by closing and reopening the window.
This is thanks to @jobisoft.
Released 1.6.2

**Comment by HugoGresse** [comment]
Thunderbird 115 and Deepl v1.6.2 works! thanks ! ❤️

**Comment by itagagaki** [comment]
Thanks everyone. I seem to have solved this issue.
